From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martin Finkel <me@martinfinkel.com>
Date: Wed, 5 Aug 2026 17:30:00 +0300
Subject: [PATCH] libdsm: cast getsockopt option value on Windows

Winsock declares the option value as char *, while POSIX accepts void *.
Newer clang rejects the implicit int * conversion when building for UCRT.
---
 .../0002-cast-getsockopt-option-value.patch   | 25 ++++++++++++++++++++
 contrib/src/libdsm/rules.mak                  |  1 +
 2 files changed, 26 insertions(+)
 create mode 100644 contrib/src/libdsm/0002-cast-getsockopt-option-value.patch

diff --git a/contrib/src/libdsm/0002-cast-getsockopt-option-value.patch b/contrib/src/libdsm/0002-cast-getsockopt-option-value.patch
new file mode 100644
--- /dev/null
+++ b/contrib/src/libdsm/0002-cast-getsockopt-option-value.patch
@@ -0,0 +1,25 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Martin Finkel <me@martinfinkel.com>
+Date: Wed, 5 Aug 2026 17:30:00 +0300
+Subject: [PATCH] netbios: cast getsockopt option value
+
+Winsock declares optval as char *, while POSIX accepts void *.
+---
+ src/netbios_session.c | 2 +-
+ 1 file changed, 1 insertion(+), 1 deletion(-)
+
+diff --git a/src/netbios_session.c b/src/netbios_session.c
+--- a/src/netbios_session.c
++++ b/src/netbios_session.c
+@@ -123,7 +123,7 @@ netbios_session_connect(struct netbios_session *s, const char *host,
+ 
+             if (FD_ISSET(s->socket, &write_fds))
+             {
+-                if (getsockopt(s->socket, SOL_SOCKET, SO_ERROR, &ret,
++                if (getsockopt(s->socket, SOL_SOCKET, SO_ERROR, (char *)&ret,
+                                &(socklen_t){ sizeof (ret) }) || ret)
+                 {
+                     errno = ret;
+-- 
+2.50.1.windows.1
+
diff --git a/contrib/src/libdsm/rules.mak b/contrib/src/libdsm/rules.mak
--- a/contrib/src/libdsm/rules.mak
+++ b/contrib/src/libdsm/rules.mak
@@ -21,6 +21,7 @@ libdsm: libdsm-$(LIBDSM_VERSION).tar.xz .sum-libdsm
 	$(UNPACK)
 	$(APPLY) $(SRC)/libdsm/0001-Avoid-relying-on-implicit-function-declarations.patch
 	$(APPLY) $(SRC)/libdsm/0001-use-GetCurrentProcessId-for-the-process-ID-on-Window.patch
+	$(APPLY) $(SRC)/libdsm/0002-cast-getsockopt-option-value.patch
 	$(MOVE)
 
 DEPS_libdsm = libtasn1 iconv
-- 
2.50.1.windows.1

